Edits

mahmutovAR opened the pull request:
документация функций не исправлена, сначала надо привести аргументы и функции в порядок


сейчас скрипт берет все файлы из initial folder и делит на группы:
- изображения (по расширению)
- не изображения (по расширению)
- изображения у которых не вычисляется размер

при копировании проверяется название файла в соответствующей папке в target folder:
- если названия совпадают и совпадают контрольные суммы, файл не копируется
- если названия совпадают, а контрольные суммы нет - переименовывается копируемый файл и проверка повторяется
    переименование выглядит так: перед старым названием добавляется !1, если такой файл существует, то !2 и т.д.

в режиме dryrun - в готовых папках указывается исходное название без проверки названий файлов

если проверка контрольных сумм после сортировки не прошла успешно, то удаление файлов из initial folder отменяется
(соответствующее сообщение появляется в консоли)


вопрос по коду:

- перемещение файлов реализовано так:
   1 копирование файлов
   2 удаление папки initial_folder
   3 создание папки initial_folder
	это сильно криво?
	python позволяет сделать как сейчас реализовано, или
        сначала удалить все файлы из всех папок, а потом уже удалить все пустые папки


вопросы по функционалу:

- если один и тот же файл содержится в разных папках с разными именами - необходимо ли удалять клона?

- нужен ли в консоли отчет после верификации? (всего __ файлов, __ изображений, __ не изображений....)

- какой лучше указать размер буфера при вычислении контрольной суммы? сейчас block_size = 65536

mahmutovAR committed:
edited version

документация функций не исправлена, сначала надо привести аргументы и функции в порядок

сейчас скрипт берет все файлы из initial folder и делит на группы:
- изображения (по расширению)
- не изображения (по расширению)
- изображения у которых не вычисляется размер

при копировании проверяется название файла в соответствующей папке в target folder:
- если названия совпадают и совпадают контрольные суммы, файл не копируется
- если названия совпадают, а контрольные суммы нет - переименовывается копируемый файл и проверка повторяется
    переименование выглядит так: перед старым названием добавляется !1, если такой файл существует, то !2 и т.д.

в режиме dryrun - в готовых папках указывается исходное название без проверки названий файлов

если проверка контрольных сумм после сортировки не прошла успешно, то удаление файлов из initial folder отменяется
(соответствующее сообщение появляется в консоли)

вопрос по коду:

- перемещение файлов реализовано так:
   1 копирование файлов
   2 удаление папки initial_folder
   3 создание папки initial_folder
	это сильно криво?
	python позволяет сделать как сейчас реализовано, или
        сначала удалить все файлы из всех папок, а потом уже удалить все пустые папки

вопросы по функционалу:

- если один и тот же файл содержится в разных папках с разными именами - необходимо ли удалять клона?

- нужен ли в консоли отчет после верификации? (всего __ файлов, __ изображений, __ не изображений....)

- какой лучше указать размер буфера при вычислении контрольной суммы? сейчас block_size = 65536

mahmutovAR committed:
Merge branch 'edits' of https://github.com/mahmutovAR/imagesort into edits

starichkovva commented:
> если один и тот же файл содержится в разных папках с разными именами - необходимо ли удалять клона?

Алгоритм проверки контрольной суммы не даёт стопроцентной гарантии идентичности файлов при идентичных хэш-суммах, я бы сказал клон нужно класть рядом добавляя суффикс к названию.

> нужен ли в консоли отчет после верификации? (всего __ файлов, __ изображений, __ не изображений....)

Почему бы и нет, это же консольная утилита получается, пусть будет.

> какой лучше указать размер буфера при вычислении контрольной суммы? сейчас block_size = 65536

Это сложный вопрос, потому как оптимальный, с точки зрения производительности, размер буфера зависит как от конкретной модели жёсткого диска или SSD, так и от текущей файловой системы на этом диске.

Можно либо оставить 64 килобайта как сейчас, либо сделать этот параметр настраиваемым, оставив значение по умолчанию в 64 килобайта.

starichkovva commented:
> перемещение файлов реализовано так:
1 копирование файлов
2 удаление папки initial_folder
3 создание папки initial_folder
это сильно криво?

Честно говоря не совсем понял из описания как же оно всё-таки реализовано. Но советую посмотреть [на вот эту библиотеку](https://docs.python.org/3.9/library/shutil.html), входящую в стандартный набор.

mahmutovAR commented:
> > перемещение файлов реализовано так:
> 1 копирование файлов
> 2 удаление папки initial_folder
> 3 создание папки initial_folder
> это сильно криво?
> 
> Честно говоря не совсем понял из описания как же оно всё-таки реализовано. Но советую посмотреть [на вот эту библиотеку](https://docs.python.org/3.9/library/shutil.html), входящую в стандартный набор.

При выборе режима "копирование" - файлы копируются, при выборе режима "по умолчанию" - файлы копируются[1], и только после успешной верификации файлов по контрольным суммам, сама папка initial folder удаляется[2] (со всем содержимым соответственно), и после этого создаётся новая папка с прежним названием[3]. Это криво потому, что зачем удалять саму папку, если можно удалить все папки и файлы внутри неё. Но вот реализовать нормальный способ сейчас можно так:
- заходишь во все папки по очереди,
- удаляются все файлы внутри,
- а уже потом удаляется все пустые папки
Поэтому я выбрал кривой, но не быстрый путь

starichkovva commented:
А зачем тебе исходная пустая папка после перемещения всех файлов?

mahmutovAR commented:
В принципе если все прошло успешно и все файлы раскиданы в другой папке - можно и удалить.
Еще два вопроса появилось:
1. После твоего уточнения: "...Алгоритм проверки контрольной суммы не даёт стопроцентной гарантии идентичности файлов при идентичных хэш-суммах..."
- если в папке "600х400" лежит картинка "wall_1.bmp" и скрипт пытается туда скопировать "wall_1.bmp" с идентичной хэш-суммой, то лучше сохранить оба варианта? Сейчас данный файл просто не будет скопирован.
2. Пожалуй стоит добавить проверку на совпадение initial folder и target folder - иначе по итогу все будет рассортировано, а потом благополучно удалено. Согласен?

starichkovva commented:
> если в папке "600х400" лежит картинка "wall_1.bmp" и скрипт пытается туда скопировать "wall_1.bmp" с идентичной хэш-суммой, то лучше сохранить оба варианта? Сейчас данный файл просто не будет скопирован.

Я бы сохранял оба, на манер Windows которая добавляет (1), (2), e.t.c. к имени файлов

starichkovva commented:
> Пожалуй стоит добавить проверку на совпадение initial folder и target folder - иначе по итогу все будет рассортировано, а потом благополучно удалено. Согласен?

Да, я думаю это будет один из основных сценариев использования - навести порядок в той же самой папке. Так что программа должна разгрести завалы удалив пустые папки ничего не потеряв.

mahmutovAR committed:
if initial_folder == target_folder

документация функций не исправлена, сначала надо привести аргументы и функции в порядок

сейчас скрипт берет все файлы из initial folder и делит на группы:
- изображения (по расширению)
 -не изображения (по расширению)
 -изображения, у которых не вычисляется размер

при копировании проверяется название файла в соответствующей папке в target folder:
- если названия совпадают - переименовывается копируемый файл, перед старым названием добавляется !1, если такой файл существует, то !2 и т.д.

в режиме dryrun - в готовых папках указывается исходное название без проверки названий файлов

если проверка контрольных сумм после сортировки не прошла успешно, то удаление файлов из initial folder отменяется
(соответствующее сообщение появляется в консоли)

вопрос по коду:
если initial folder и target folder совпадают, то код выполняется так:
1 создается временная папка f’{initial_folder}+temp1’ c проверкой на наличие такой папки в каталоге, если такая папка уже существует, то название будет f’{initial_folder}+temp2’ и т.д
2 во временную папку сортируются все исходные файлы
3 оригинальная папка {initial_folder} удаляется, а временная переименовывается в {initial_folder}
	это нормально?

starichkovva commented:
Давай всё-таки займемся оформлением этого проекта в GitHub:

- Завести Project на вкладке Projects как описано [в официальной доке](https://docs.github.com/en/issues/organizing-your-work-with-project-boards/managing-project-boards/about-project-boards) GitHub-а
- Создать под новый проект команду
- Назначить в команду меня и @zaharov-k

starichkovva commented on imagesort.py:
```diff
@@ -10,28 +10,26 @@
     image_types = ['.bmp', '.gif', '.png', '.jpg', '.jpeg', '.tiff', '.raw']
-    initial_folder, target_folder, remove_status, dry_run, all_files, images, not_images_status, not_images = input_parameters(image_types)
-    images_by_resolutions, not_sized_status, not_sized_files = structure_of_target_folder(initial_folder, images)
+    initial_folder, target_folder, folder_rename_status, remove_status, dry_run, all_files, images, not_images_status, not_images = parse_arguments(image_types)
```
Давай добавим следующую обёртку - если число аргументов неправильное или недостаточное для продолжения работы - печатать на экран справочную информацию о том, как правильно использовать этот скрипт.

Обычно принято добавлять специальные аргументы вида `-h` или `--help` для вывода справочной информации.

starichkovva commented on imagesort.py:
```diff
@@ -141,108 +167,95 @@ def dry_run_report(dir_path: str, target_folder: str, images_by_resolutions: dic
-        print('Verification completed with error!')
+        res_files_checksums.extend(get_checksum_from_named_folder(target_folder, 'Error files', not_sized_files, exception_dict))
+        
+    not_verified = True
```
Это объявление переменной как будто бы лишнее.

starichkovva commented on imagesort.py:
```diff
@@ -141,108 +167,95 @@ def dry_run_report(dir_path: str, target_folder: str, images_by_resolutions: dic
+
+    if not_verified:
+        print('Attention! Checksum verification completed with an error. Deleting of the initial files canceled.')
+        remove_status = False
```
Как и вот это - не вижу чтобы эта переменная использовалась в коде ниже.

starichkovva commented on imagesort.py:
```diff
@@ -141,108 +167,95 @@ def dry_run_report(dir_path: str, target_folder: str, images_by_resolutions: dic
+   
+    total_not_images=0
+    for k in not_images.keys():
+        total_not_images+=len(not_images[k])
```
Много проблем с форматированием кода - нет пробелов в выражении между переменной и операцией. Рекомендую вызвать `Code > Reformat Code` в PyCharm для автоматического исправления подобных моментов.

starichkovva commented on imagesort.py:
```diff
@@ -254,24 +267,51 @@ def getting_checksum(arg_file: str) -> str:
+        return file_name
+
+
+def get_dirs_and_files(folder: str) -> dict:
```
Хотелось бы видеть документацию на все методы.

starichkovva commented on README.md:
```diff
@@ -20,7 +21,6 @@ Files for which resolution couldn't be determined will be copied or moved to the
 * copy - sorts images and copies them to the target folder
```
```
$ python3 imagesort.py ~/dev/walls/ ~/dev/wallpapers dryrun
Error! Entered target folder doesn't exist: /home/starichkov/dev/wallpapers
```
Было бы удобно, если бы скрипт создавал target folder, а не бил по рукам. Мы всё-таки делаем задачу по автоматизации действий.

starichkovva commented on README.md:
```diff
@@ -20,7 +21,6 @@ Files for which resolution couldn't be determined will be copied or moved to the
 * copy - sorts images and copies them to the target folder
 
```
Если по умолчанию скрипт вызывается в режиме перемещения файлов, то рекомендую параметр для выбора режима делать опциональным. Если не указан - default, если указан - то использовать тот режим что указали.

Иначе получается странная запись вида:

```
$ python3 imagesort.py ~/dev/walls2/ ~/dev/walls2/ default
```

где `default` в конце явно не очень смотрится.

starichkovva commented on imagesort.py:
```diff
@@ -10,28 +10,26 @@
-                              not_images, not_sized_status, not_sized_files, exception_dict)
+            remove_initial_files(initial_folder, target_folder)
+        if folder_rename_status:
+            rename_target_folder(initial_folder, target_folder, dir_path)
```
Если указать скрипту одну и ту же папку как исходную и конечную, скрипт в конце умирает:

```
$ python3 imagesort.py ~/dev/walls2/ ~/dev/walls2/ default
Input parameters are:
Initial folder: /home/***/dev/walls2/
Target folder:  /home/***/dev/walls2/-temp1
Additional option:      sort image and move from the initial folder to the target folder

Information! There is(are) "not images" file(s) in the initial folder so directory "Other files" will be created.
Checksum verification completed successfully
"ImageSort" report:
Total files in the initial folder               938
Images in the initial folder                    919
Not images in the initial folder                19
Not sized images in the initial folder          0
Traceback (most recent call last):
  File "/home/***/imagesort.py", line 317, in <module>
    main()
  File "/home/***/imagesort.py", line 29, in main
    rename_target_folder(initial_folder, target_folder, dir_path)
  File "/home/***/imagesort.py", line 255, in rename_target_folder
    os.rename(target_folder, initial_folder)
FileNotFoundError: [Errno 2] No such file or directory: '/home/***/dev/walls2/-temp1' -> '/home/***/dev/walls2/'```

mahmutovAR committed:
script, *args = sys.argv

imagesort.py -dryrun "initial folder"
imagesort.py -copy "initial folder" "target folder"
imagesort.py -remove "initial folder" "target folder"
imagesort.py "initial folder"
imagesort.py -help

mahmutovAR committed:
with argparse

argparse is used for parsing args
--help mode was added

mahmutovAR committed:
change name function and README were  modified

mahmutovAR committed:
change name function and description were modified

mahmutovAR committed:
fix bugs with not_images ant not_sized_images dictionaries

mahmutovAR committed:
final version of "ImageSort"

mahmutovAR commented on imagesort.py:
```diff
@@ -1,249 +1,227 @@
+                                          for k, v in images_attributes.items()
+                                          if v[1] == resolution]
+                             for resolution in set([attributes[1]
+                                                    for attributes in images_attributes.values()])}
```
заменил генератор словаря на циклы

mahmutovAR commented on imagesort.py:
```diff
@@ -1,249 +1,227 @@
+
+    output_files = {f'{resolution}': [images_attributes[image][0]
+                                      for image in images_by_resolutions[resolution]]
+                    for resolution in images_by_resolutions.keys()}
```
заменил генератор словаря на циклы

mahmutovAR commented on imagesort.py:
```diff
@@ -1,249 +1,227 @@
         output_files['Error files'] = not_sized_files
-    sorted_output_files = {k: output_files[k]
+
+    sorted_output_files = {k: sorted(output_files[k])
```
заменил генератор словаря на циклы

mahmutovAR committed:
dict generators and images formats

starichkovva commented:
Давай смёржим это всё в мастер, чтобы была одна актуальная ветка, по которой я отмечу уже code style моменты.
